MailDev custom resource tutorial. (#755)
dotnet/docs-aspire: docs/extensibility/custom-resources.md

```csharp
app.MapGet("/subscribe", async (SmtpClient smtpClient, string email) => {
    var message = new MailMessage("newsletter@yourcompany.com", email);
    message.Subject = "Welcome to our newsletter!";
    message.Body = "Thank you for subscribing to our newsletter!";
    await smtpClient.SendMailAsync(message);
});

app.MapGet("/unsubscribe", async (SmtpClient smtpClient, string email) => {
    var message = new MailMessage("newsletter@yourcompany.com", email);
    message.Subject = "You are unsubscribed from our newsletter!";
    message.Body = "Sorry to see you go. We hope you will come back soon!";
    await smtpClient.SendMailAsync(message);
});
```

Once the `Program.cs` file is updated launch the AppHost and use your browser, or `curl` to hit the following URLs:

```bash
# Remember to replace the port to match your local environment.
curl https://localhost:7289/subscribe?email=test@test.com
curl https://localhost:7289/unsubscribe?email=test@test.com
```

If those API calls return a successful response then you should be able to click on the `maildev` resource the dashboard and the MailDev UI will show the emails that have been sent to the SMTP endpoint.

> ![E-mails visible in the MailDev UI](./media/maildev-emails.png)

Reviewer: I think we need an explanation of what we did and the pieces after the sample or maybe explained after each section. It would also be nice to show the manifest representation of this resource.

Author: Expanded the technical details. Let me know if this covers what you are after. If we want more I think it needs to be a separate article because it can be pretty overwhelming.

Reviewer: It looks good.

```dotnetcli
dotnet run --project MailDevResource.AppHost/MailDevResource.AppHost.csproj
```

The console output should look similar to the following:

```dotnetcli
Building...
info: Aspire.Hosting.DistributedApplication[0]
      Aspire version: 8.0.0-preview.7.24224.7+147c1d2cccdff3b185517a85f095ba590608500c
info: Aspire.Hosting.DistributedApplication[0]
      Distributed application starting.
info: Aspire.Hosting.DistributedApplication[0]
      Application host directory is: C:\Code\MailDevTutorial\MailDevResource\MailDevResource.AppHost
info: Aspire.Hosting.DistributedApplication[0]
      Now listening on: https://localhost:17297
info: Aspire.Hosting.DistributedApplication[0]
      Login to the dashboard at https://localhost:17297/login?t=f6104b597046542abe63c6d7cb1a7709
info: Aspire.Hosting.DistributedApplication[0]
      Distributed application started. Press Ctrl+C to shut down.
```

Open the dashboard link in the browser and the dashboard:

![Empty dashboard for test project](./media/maildevresource-empty-dashboard.png)

Press `CTRL-C` to shutdown the application (you can close the browser tab).

### Step 2: Creating library for resource extension

> .NET Aspire resources are just classes and methods contained within a class library that references the .NET Aspire Hosing library (`Aspire.Hosting`). By placing the resource in a separate project you can more easily share it between .NET Aspire-based applications and potentially package it and share it on NuGet.

Reviewer 2: ```suggestion
.NET Aspire resources are just classes and methods contained within a class library that references the .NET Aspire Hosting library (`Aspire.Hosting`). By placing the resource in a separate project you can more easily share it between .NET Aspire-based applications and potentially package it and share it on NuGet.
```

![Environment variables for Newsletter Service in .NET Aspire Dashboard](./media/maildev-envvar.png)

### Step 4b: Use connection string to send message.

To use the SMTP connection details that were injected into the newsletter service project we will inject an instance of <xref:System.Net.Mail.SmtpClient> into the dependency injection container as a singleton. Add the following code to the `Program.cs` file in the `MailDevResource.NewsletterService` project to setup the singleton service.

```csharp
builder.Services.AddSingleton<SmtpClient>((sp) =>
{
    var smtpUri = new Uri(builder.Configuration.GetConnectionString("maildev")!);
    var smtpClient = new SmtpClient(smtpUri.Host, smtpUri.Port);
    return smtpClient;
});
```

To test the client we will add two simple subscribe and unsubscribe GET methods to the newsletter service. Add the following code to the `Program.cs` file in the `MailDevResource.NewsletterService` project to setup the ASP.NET Core routes:

```csharp

> app.MapGet("/subscribe", async (SmtpClient smtpClient, string email) => {
>     var message = new MailMessage("newsletter@yourcompany.com", email);
>     message.Subject = "Welcome to our newsletter!";
>     message.Body = "Thank you for subscribing to our newsletter!";
>     await smtpClient.SendMailAsync(message);
> });
> 
> app.MapGet("/unsubscribe", async (SmtpClient smtpClient, string email) => {
>     var message = new MailMessage("newsletter@yourcompany.com", email);
>     message.Subject = "You are unsubscribed from our newsletter!";
>     message.Body = "Sorry to see you go. We hope you will come back soon!";
>     await smtpClient.SendMailAsync(message);
> });

Reviewer 2: If I recall correctly, the `MailMessage` type is disposable, no? Let's slap a `using` on that, and fix the opening curly brace.

```suggestion
app.MapGet("/subscribe", async (SmtpClient smtpClient, string email) =>
{
    using var message = new MailMessage("newsletter@yourcompany.com", email);
    message.Subject = "Welcome to our newsletter!";
    message.Body = "Thank you for subscribing to our newsletter!";
    await smtpClient.SendMailAsync(message);
});

app.MapGet("/unsubscribe", async (SmtpClient smtpClient, string email) =>
{
    using var message = new MailMessage("newsletter@yourcompany.com", email);
    message.Subject = "You are unsubscribed from our newsletter!";
    message.Body = "Sorry to see you go. We hope you will come back soon!";
    
    await smtpClient.SendMailAsync(message);
});
```

```

> Once the `Program.cs` file is updated launch the AppHost and use your browser, or `curl` to hit the following URLs:

Reviewer 2: ```suggestion
Once the _Program.cs_ file is updated launch the app host and use your browser, or `curl` to hit the following URLs:
```

```bash
# Remember to replace the port to match your local environment.
curl https://localhost:7289/subscribe?email=test@test.com
curl https://localhost:7289/unsubscribe?email=test@test.com
```

If those API calls return a successful response then you should be able to click on the `maildev` resource the dashboard and the MailDev UI will show the emails that have been sent to the SMTP endpoint.

![E-mails visible in the MailDev UI](./media/maildev-emails.png)

## Technical details

> ### IValueProvider and IManifestExpressionProvider

Reviewer 2: ```suggestion
### `IValueProvider` and `IManifestExpressionProvider`
```

> ### IValueProvider and IManifestExpressionProvider

Reviewer: ```suggestion
### ReferenceExpression
```

I would focus this part of `ReferenceExpression` since you use that type directly. Then we can write a more detailed piece on the underlying interfaces.

Author: Is this more about the title of the section or did you want content changes. I do mention the interfaces but don't go into a whole heck of a lot of detail.

The fundamental problem that these types help solve is deferring resolution of concrete configuration information until _all_ of the information is available.

For example, in the `MailDevResource` example we expose a property called `ConnectionStringExpression` as required by the <xref:Aspire.Hosting.ApplicationModel.IResourceWithConnectionString> interface. The type of the property is <xref:Aspire.Hosting.ApplicationModel.ReferenceExpression> and is created by passing in an interpolated string to the <xref:Aspire.Hosting.ApplicationModel.ReferenceExpression.Create> method.

```csharp
public ReferenceExpression ConnectionStringExpression => ReferenceExpression.Create(
        $"smtp://{SmtpEndpoint.Property(EndpointProperty.Host)}:{SmtpEndpoint.Property(EndpointProperty.Port)}"
        );
```

> The signature for the <xref:Aspire.Hosting.ApplicationModel.ReferenceExpression.Create> method is as follows:

Reviewer 2: @mitchdenny - you asked why these were wrong, when you see the published API docs that correspond to this xref, you can verify the name matches the [HTML's](https://learn.microsoft.com/en-us/dotnet/api/aspire.hosting.applicationmodel.referenceexpression.create?view=dotnet-aspire-8.0) `content` attribute value of the <head><meta name="ms.assetid"></meta></head>`. In this case, you'd see this:

```html
<meta name="ms.assetid" content="Aspire.Hosting.ApplicationModel.ReferenceExpression.Create*">
```

And that `*` was what you're missing, `%2A`.

```suggestion
The signature for the <xref:Aspire.Hosting.ApplicationModel.ReferenceExpression.Create%2A> method is as follows:
```

```csharp
public static ReferenceExpression Create(in ExpressionInterpolatedStringHandler handler)
```

> Notice that this is not a regular <xref:System.String> argument. Method makes use of the [interpolated string handler pattern](../../csharp/whats-new/tutorials/interpolated-string-handler) in .NET to capture the interpolated string template and the values referenced within it to allow for custom processing. In the case of .NET Aspire we capture these details in a <xref:Aspire.Hosting.ApplicationModel.ReferenceExpression> which can be evaluated as each value referenced in the interpolated string becomes available.

Reviewer 2: The reason that this `csharp` doesn't work, is that you're trying to use relative path resolution on a file that's not found in this docset. Instead, you'd want to use site relative path resolution. This has all the details, if you're interested in learning about how it's done: https://review.learn.microsoft.com/en-us/help/platform/links-how-to?branch=main

```suggestion
Notice that this is not a regular <xref:System.String> argument. Method makes use of the [interpolated string handler pattern](/dotnet/csharp/whats-new/tutorials/interpolated-string-handler) in .NET to capture the interpolated string template and the values referenced within it to allow for custom processing. In the case of .NET Aspire we capture these details in a <xref:Aspire.Hosting.ApplicationModel.ReferenceExpression> which can be evaluated as each value referenced in the interpolated string becomes available.
```

The `MailDevResource` automatically gets included in the manifest because it is derived from <xref:Aspire.Hosting.ApplicationModel.ContainerResource>. Resource authors can choose to suppress outputting content to the manifest by overriding the default behavior using the following code (inside the `AddMailDev(...)` extension)

```csharp
public static IResourceBuilder<MailDevResource> AddMailDev(this IDistributedApplicationBuilder builder, string name, int? httpPort = null, int? smtpPort = null)
{
    var resource = new MailDevResource(name);
    return builder.AddResource(resource)
                    .WithImage(MailDevContainerImageTags.Image)
                    .WithImageRegistry(MailDevContainerImageTags.Registry)
                    .WithImageTag(MailDevContainerImageTags.Tag)
                    .WithHttpEndpoint(targetPort: 1080, port: httpPort, name: MailDevResource.HttpEndpointName)
                    .WithEndpoint(targetPort: 1025, port: smtpPort, name: MailDevResource.SmtpEndpointName)

>                     .WithAnnotation(ManifestPublishingCallbackAnnotation.Ignore); // This line added!

Reviewer: Why this an not ExcludeFromManifest?

Author: Would you believe that I keep forgetting it exists. Fixed.

info: Aspire.Hosting.DistributedApplication[0]
      Distributed application started. Press Ctrl+C to shut down.
```

Select the [dashboard link in the browser](../fundamentals/dashboard/explore.md#dashboard-authentication) to see the .NET Aspire dashboard:

:::image type="content" source="media/maildevresource-empty-dashboard.png" lightbox="media/maildevresource-empty-dashboard.png" alt-text="A screenshot of the empty .NET Aspire dashboard for test project.":::

Press <kbd>Ctrl</kbd>+<kbd>C</kbd> to shutdown the app (you can close the browser tab).

## Create library for resource extension

.NET Aspire resources are just classes and methods contained within a class library that references the .NET Aspire Hosting library (`Aspire.Hosting`). By placing the resource in a separate project you can more easily share it between .NET Aspire-based apps and potentially package and share it on NuGet.

1. Create the class library project named _MailDev.Hosting_.

    ```dotnetcli
    dotnet new classlib -o MailDev.Hosting
    ```

1. Add `Aspire.Hosting` to the class library as a package reference.

    ```dotnetcli

>     dotnet add ./MailDev.Hosting\MailDev.Hosting.csproj package Aspire.Hosting --version 8.0.0

Reviewer 3: ```suggestion
    dotnet add ./MailDev.Hosting/MailDev.Hosting.csproj package Aspire.Hosting --version 8.0.0
```
